bart/sched/pelt.py

```diff
@@ -0,0 +1,377 @@
+
+    The model allows to compute statistics on the generated signal, over a
+    specified time frame. It provides also an embedded support to plot the
+    generated signal along with a set of expected thresholds and values
+    (e.g. half-life, min/max stability values)
+    """
+
+    _sample_us = 1024
+    _signal_max = 1024
+
+    def __init__(self, init_util=0, half_life_ms=32, decay_cap_ms=None):
+        """
+        Initialize the PELT simulator for a specified task.
+
+        :param init_util: the initial utilization for the task
+        :type  init_util: int
+
+        :param half_life_ms: the [ms] interval required by a signal to
+                             increased/decreased by half of its range
+        :type  half_life_ms: int
+
+        :param decay_cap_ms: the [ms] interval after which we do not decay
+                             further a signal
+        :type  decay_cap_ms: int
+
+        """
+        self.init_util = init_util
+        self.half_life_ms = half_life_ms
+        self.decay_cap_ms = decay_cap_ms
+
+        self._geom_y = pow(0.5, 1./half_life_ms)
+        self._geom_u = float(Simulator._signal_max) * (1. - self._geom_y)
+
+        print "PELT Configured with   :"
+        print "  initial utilization  : ", self.init_util
+        print "  half life (HL)  [ms] : ", self.half_life_ms
+        print "  decay capping @ [ms] : ", self.decay_cap_ms
+        print "  y =    0.5^(1/HL)    : {:6.3f}".format(self._geom_y)
+        print "  u = {:5d}*(1 - y)    : {:6.3f}".format(Simulator._signal_max, self._geom_u)
+
+        self._df = None
+
+    def _geom_sum(self, u_1, active_us=0):
+        """
+        Geometric add the specified value to the u_1 series.
+
+        NOTE: the current implementation assume that the task was active
+              (active_us != 0) or sleeping (active_us == 0) a full PELT
+              sampling interval (by default 1024[us])
+        """
+        # Decay previous samples
+        u_1 *= self._geom_y
+        # Add current sample (if task was active)
+        if active_us:
+            u_1 += self._geom_u
+        return u_1
+
+    def stable_range(self, task):
+        """
+        Compute the stability ranges for the specified (periodic) task
+
+        Here we use:
+        - Decay factor:       y = 0.5^(1/half_life)
+        - Max stable value:       (1-y^r) / (1-y^p)
+        - Min stable value: y^i * (1-y^r) / (1-y^p)
+
+        :param task: the task we want to simulate the PELT signal for, by
+                     default is the task used to intialize the Simulator
+                     (when specified)
+        :type  task: Task
+
+        :return: A tuple representing the minimum and maximum value of the PELT
+                 singal once stable.
+
+        """
+
+        # Validate input parameters
+        if not isinstance(task, Task):
+            raise ValueError("Wrong time for task parameter")
```

Comment:
s/time/type/? Anyway, I don't think we should check types of parameters - if it looks like a Task, talks like a Task and has all the attributes of a Task let's just call it a Task